src/daemon/daemon.cpp

```diff
@@ -1544,119 +1722,72 @@ try // clang-format on
     mpl::ClientLogger<InfoReply, InfoRequest> logger{mpl::level_from(request->verbosity_level()), *config->logger,
                                                      server};
     InfoReply response;
+    InstanceSnapshotsMap instance_snapshots_map;
     bool have_mounts = false;
     bool deleted = false;
-    auto fetch_info = [&](VirtualMachine& vm) {
-        const auto& name = vm.vm_name;
-        auto info = response.add_info();
-        auto present_state = vm.current_state();
-        info->set_name(name);
-        if (deleted)
-        {
-            info->mutable_instance_status()->set_status(mp::InstanceStatus::DELETED);
-        }
-        else
-        {
-            info->mutable_instance_status()->set_status(grpc_instance_status_for(present_state));
-        }
+    bool snapshots_only = request->snapshots();
+    response.set_snapshots(snapshots_only);
 
-        auto vm_image = fetch_image_for(name, config->factory->fetch_type(), *config->vault);
-        auto original_release = vm_image.original_release;
+    auto populate_info = [&](VirtualMachine& vm, const std::shared_ptr<const Snapshot>& snapshot) {
+        auto* details = response.add_details();
+        if (snapshot)
+            populate_snapshot_info(vm, snapshot, details, have_mounts); // TODO@snapshots remove have_mounts
+        else
+            populate_instance_info(vm, details, request->no_runtime_information(), deleted, have_mounts);
+    };
```

Comment:
OK, let me try to address each of these remarks separately.

> First populate_snapshot_info is an anonymous local function yet populate_instance_info is a Daemon member function.

`VirtualMachine` objects do not have access to all the information that pertain to them, so `populate_instance_info` needs the daemon for things like mounts and specs. With snapshots we tried to be more object-oriented, so you only need the `vm` and `snapshot` objects to fill all the information that is required. 

I can't think of an appropriate alternative that would reintroduce the symmetry without breaking something else: the method wouldn't work as a free function and the free function shouldn't be a method that doesn't use this (we could make it a private static method, but I don't see the gain). Wrapping wouldn't help much and lambdas would reduce encapsulation unnecessarily.

> Also, it's capturing all the variables which seems a bit unnecessary.

OK, this can be improved, especially because we're also capturing this and that is not all that obvious. I will do that

> Also, it looks to me that populate_info is called under conditions that already determine if there are snapshots or not, so this seems like double logic to me. I would think the populate_snapshot_info() and populate_instance_info() could be called directly where appropriate below based on the logic already there.

Yes, it is not ideal. The goal was to make sure that we always called `add_details()` where appropriate, without repeating multiple times (this was a source of bugs in the past). It also avoids having to repeat the same parameter list multiple times. But I recognize it looks like we're going back and forth on deciding what info to actually fill. I will see what I can do about this.

Reply:
Ok, I understand the asymmetry of where the two functions live.  But yes, the other two things here you address I do think need a little more work.  Thanks!

Reply:
Much, much better, thank you!!!!